src/Corax/Querying/TermsReader.cs

```diff
@@ -41,6 +51,79 @@ public string GetTermFor(long id)
         return s;
     }
     
+    /// <summary>
+    /// Read terms in bulk. The result is not associated with ids by index.
+    /// Caution:
+    /// - the `termsSet` span is being invalidated between calls.
+    /// - the method modifies 'ids' (change the order) of the buffer (!)
+    /// - the order of the terms is based on sorted ids values
+    /// </summary>
+    public int GetAllTermsFromSet(Span<long> ids, out Span<UnmanagedSpan> termsSet)
+    {
+        const int pageSizeShift = 13;
+        Debug.Assert(1 << pageSizeShift == (long)Voron.Global.Constants.Storage.PageSize);
+        var maxToProcess = Math.Min(ids.Length, 1024);
+        ids = ids[..maxToProcess];
+
+        if (ids.IsEmpty)
+        {
+            termsSet = default;
+            return 0;
+        }
+        
+        _termsLocation.ResetAndEnsureCapacity(ids.Length);
+        _termsLocation.Count = ids.Length;
+        _lookup.GetFor(ids, _termsLocation.ToSpan(), -1L);
+        
+        var idX = 0;
+        _pagesToPrefetch.ResetAndEnsureCapacity(ids.Length);
+        _pagesToPrefetch.Count = ids.Length;
+        _termsLocation.CopyTo(_pagesToPrefetch.ToSpan(), startFrom: 0);
+        
+        if (AdvInstructionSet.IsAcceleratedVector512)
+        {
+            var N = Vector512<long>.Count;
+            for (; idX + N < _pagesToPrefetch.Count; idX += N)
+            {
+                var ptr = _pagesToPrefetch.RawItems + idX;
+                var containers = Vector512.Load(ptr);
+                Vector512.ShiftRightArithmetic(containers, pageSizeShift).Store(ptr);
+            }
+        }
+        else if (AdvInstructionSet.IsAcceleratedVector256)
+        {
+            var N = Vector256<long>.Count;
+            for (; idX + N < _pagesToPrefetch.Count; idX += N)
+            {
+                var ptr = _pagesToPrefetch.RawItems + idX;
+                var containers = Vector256.Load(ptr);
+                Vector256.ShiftRightArithmetic(containers, pageSizeShift).Store(ptr);
+            }
+        }
+        else if (AdvInstructionSet.IsAcceleratedVector128)
+        {
+            var N = Vector128<long>.Count;
+            for (; idX + N < _pagesToPrefetch.Count; idX += N)
+            {
+                var ptr = _pagesToPrefetch.RawItems + idX;
+                var containers = Vector128.Load(ptr);
+                Vector128.ShiftRightArithmetic(containers, pageSizeShift).Store(ptr);
+            }
+        }
+
```

Comment:
Did you measure differences here between using 512 bits fallbacks instead of having the code replicated for 256 and 128? As funny as it sounds there are cases that using the fallback is faster than having all the specific code (Microsoft made that mistake with CosineSimilarity).

Reply:
was this checked?

Reply:
Based on those numbers, I would just do 512.

Reply:
And when you don't have 512 (very common) ?

What I don't understand, after JIT, I would expect both function to be exactly the same.
Isn't `AdvInstructionSet.IsAcceleratedVector512` a jit time constant?

Reply:
@redknightlois ?

Reply:
I think there is a disconnection in what I mean and what @ayende is asking... 

I would write this code only:

```
            var N = Vector512<long>.Count;
            for (; idX + N < _pagesToPrefetch.Count; idX += N)
            {
                var ptr = _pagesToPrefetch.RawItems + idX;
                var containers = Vector512.Load(ptr);
                Vector512.ShiftRightArithmetic(containers, pageSizeShift).Store(ptr);
            }
```

No checks... just plain code... why? Because that will fallback through the execution based on what length of instructions we have. In 256 it will issue 2x256 instead of a single 512. In 128 it will issue 4x128 instead of a single 512. Which is essentially free unrolling.

Reply:
@maciejaszyk as I see this comment isn't still resolved. Can you please take a look here?